src/java.desktop/share/classes/sun/swing/FilePane.java

```diff
@@ -1189,19 +1191,28 @@ public Component getTableCellRendererComponent(JTable table, Object value,
                 Icon icon = chooser.getIcon(file);
                 setIcon(icon);
 
-            } else if (value instanceof Long) {
-                long len = ((Long) value) / 1024L;
+            } else if (value instanceof Long len) {
                 if (listViewWindowsStyle) {
-                    text = MessageFormat.format(kiloByteString, len + 1);
+                    if(len == 0) {
+                        text = MessageFormat.format(byteString, len);
+                    } else {
+                        len /= 1024L;
+                        text = MessageFormat.format(kiloByteString, len + 1);
+                    }
                 } else if (len < 1024L) {
-                    text = MessageFormat.format(kiloByteString, (len == 0L) ? 1L : len);
+                    text = MessageFormat.format(byteString, len);
```

Comment:
For len < 1024 say 433, does it now show 433 bytes or 433 B? 
Since in resource file, you are using {0} B and not {0} bytes..I guess in ubuntu native, it shows 433 bytes and not 433 B..

Reply:
It will show 433B now, in ubuntu native it shows 433 bytes. 
bytes replaced by B to handle plural forms, otherwise for 1 byte file it show 1 bytes.

Reply:
i guess it's not right then, no? if you are changing to mimic native, then it should be "bytes" and not "B". Just for 1 byte case, you cannot compromise on other file size(s), in my opinion.

Reply:
In that case to handle 1 byte, do i need to add separate property in properties file?
If filesize=1 then can get "byte" string from properties file and can format using MessageFormat?

Reply:
am not sure if resource file can support {1} byte case...if that is supported, then you can use that...but likelihood of 1 byte file is very rare so I guess we can ignore it, that's my opinion.

Reply:
ok, should I revert back "B" to "bytes" in basic.properties and commit?

Reply:
I would like to emphasize that this issue is complicated.  Plural rules vary from language to language, adding plural strings to resource bundles adds costs in terms of translation and testing.  

If we were to support proper plurals, one may use ChoiceFormat, though it seems to fail for Polish, see
https://stuartgunter.wordpress.com/2011/08/09/java-i18n-pluralisation-using-choiceformat/

Another solution is to pull in ICU4J library, see
https://stuartgunter.wordpress.com/2011/08/14/even-better-java-i18n-pluralisation-using-icu4j/

Yet another solution is to drop the words altogether and simply show a number (of bytes in this case).  Alternatively, use a common suffix such as "B", but this may not look standard.

Reply:
another suggestion: to avoid locale-dependent complexity simply show the number of bytes as a number if length < 1000.  this will also avoid involving the group separator - as far as I know, no language on earth uses group separator for numbers in the range 0...999

0 - "0"
1 - "1"
999 - "999"
1000 - "1 KB"
etc.

Reply:
ChoiceFormat has been used now to handle formatting of file size in plural forms. For file size either 0 or more than 1 byte it shows <file length> bytes else 1 byte. 
@naotoj @andy-goryachev-oracle Please check if it is correct.

Reply:
Upon further consideration, and after discussing the state of plural forms support in JDK with @naotoj I realized that we cannot use ChoiceFormat here at all.  The numeric choices, and their corresponding resource strings are locale-specific, and in some cases simply cannot be coded as a finite array (see Arabic example above).

One possible solution would be to add proper support for plural forms to MessageFormat (which is definitely out of scope for this bug), or perhaps - if showing sizes 0...999 as numbers is not acceptable - to simply fix the 0 bytes case using KB format.  That is,

0 - "0 KB"
1  - "1 KB"
...

(i.e. handle the 0 bytes case separately):

                long len = ((Long) value);

if(len == 0) {
  // shows "0 KB" for 0 bytes
  text = MessageFormat.format(kiloByteString, 0);
} else {
  // retain existing logic
  len /= 1024L;

                if (listViewWindowsStyle) {
                    text = MessageFormat.format(kiloByteString, len + 1);
                } else if (len < 1024L) {
                    text = MessageFormat.format(kiloByteString, (len == 0L) ? 1L : len);
                } else {
                    len /= 1024L;
                    if (len < 1024L) {
                        text = MessageFormat.format(megaByteString, len);
                    } else {
                        len /= 1024L;
                        text = MessageFormat.format(gigaByteString, len);
                    }
                }

and revert the new resource string.

Reply:
Actually, my point was that since Swing is localized only in languages that can be dealt with (0, 1, other) choice patterns (or no plural pattern such as in Japanese), `ChoiceFormat` can still be used. Once we support complex plurals such as Arabic, we may need a more sophisticated solution.

Reply:
We seem to have resources localized in French.  French rules are as follows:
one → n within 0..2 and n is not 2;
other → everything else

ChoiceFormat simply cannot be used for number localization, we should remove it being mentioned in MessageFormat javadoc, at least remove the example with plurals.

The whole idea is that localization process does not require a code change.  With ChoiceFormat, it does - as the choices depend on language, and rules for some languages cannot be expressed as a set of choice, as for example in Ukrainian and many other Slavic languages:

one → n mod 10 is 1 and n mod 100 is not 11;
few → n mod 10 in 2..4 and n mod 100 not in 12..14;
many → n mod 10 is 0 or n mod 10 in 5..9 or n mod 100 in 11..14;
other → everything else

Reply:
Yes, you are right that `ChoiceFormat` is not for general localization. However, the supported UI translations in the JDK are English, German (new in 19), Japanese, and Simplified Chinese. (https://www.oracle.com/java/technologies/javase/jdk18-suported-locales.html for JDK18). That's why I made the point.

Reply:
I have reverted back the changes for basic.properties file and FileChooser now shows the file size as below:

0 - "0 KB"
1 - "1 KB" and so on...